src/main/scala/uncore/devices/Plic.scala

```diff
@@ -182,14 +176,28 @@ class TLPLIC(params: PLICParams)(implicit p: Parameters) extends LazyModule
       PLICConsts.enableBase(i) -> e.map(b => RegField(1, b))
     }
 
+    val claimer = Wire(init = Vec.fill(nHarts){Bool(false)})
+    val claiming = Wire(init = Vec.tabulate(nHarts){i => Mux(claimer(i), UIntToOH(maxDevs(i)), UInt(0, width=log2Up(nDevices+1)))})
+    val claimedDevs = Wire(init = Vec(claiming.reduceLeft( _ | _ ).toBools))
+
+    for ((pg, c) <- (pending zip gateways) zip claimedDevs) {
+      val p = pg._1
+      val g = pg._2
+      g.ready := !p
+      g.complete := false
+      when(c) {
+        p := false
+      }.elsewhen (g.valid) {
+        p := true
+      }
```

Comment:
Given that the goal is to reduce line count, it's worth seeing if this does better:

    when (c || g.valid) { p := !c }